lib/class-wp-rest-widget-updater-controller.php

```diff
@@ -0,0 +1,186 @@
+<?php
+/**
+ * Start: Include for phase 2
+ * Widget Updater REST API: WP_REST_Widget_Updater_Controller class
+ *
+ * @package gutenberg
+ * @since 5.2.0
+ */
+
+/**
+ * Controller which provides REST endpoint for updating a widget.
+ *
+ * @since 5.2.0
+ *
+ * @see WP_REST_Controller
+ */
+class WP_REST_Widget_Updater_Controller extends WP_REST_Controller {
+
+	/**
+	 * Constructs the controller.
+	 *
+	 * @access public
+	 */
+	public function __construct() {
+		$this->namespace = 'wp/v2';
+		$this->rest_base = 'widgets';
+	}
+
+	/**
+	 * Registers the necessary REST API route.
+	 *
+	 * @access public
+	 */
+	public function register_routes() {
+			register_rest_route(
+				$this->namespace,
+				// Regex representing a PHP class extracted from http://php.net/manual/en/language.oop5.basic.php.
+				'/' . $this->rest_base . '/(?P<identifier>[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*)/',
+				array(
+					'args' => array(
+						'identifier' => array(
+							'description' => __( 'Class name of the widget.', 'gutenberg' ),
+							'type'        => 'string',
+						),
+					),
+					array(
+						'methods'             => WP_REST_Server::EDITABLE,
+						'permission_callback' => array( $this, 'compute_new_widget_permissions_check' ),
+						'callback'            => array( $this, 'compute_new_widget' ),
+					),
+				)
+			);
+	}
+
+	/**
+	 * Checks if the user has permissions to make the request.
+	 *
+	 * @since 5.2.0
+	 * @access public
+	 *
+	 * @return true|WP_Error True if the request has read access, WP_Error object otherwise.
+	 */
+	public function compute_new_widget_permissions_check() {
+		// Verify if the current user has edit_theme_options capability.
+		// This capability is required to access the widgets screen.
+		if ( ! current_user_can( 'edit_theme_options' ) ) {
+			return new WP_Error(
+				'widgets_cannot_access',
+				__( 'Sorry, you are not allowed to access widgets on this site.', 'gutenberg' ),
+				array(
+					'status' => rest_authorization_required_code(),
+				)
+			);
+		}
+		return true;
+	}
+
+	/**
+	 * Returns the new widget instance and the form that represents it.
+	 *
+	 * @since 5.2.0
+	 * @access public
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function compute_new_widget( $request ) {
+		$url_params = $request->get_url_params();
+
+		$widget = $request->get_param( 'identifier' );
+
+		global $wp_widget_factory;
+
+		if (
+			null === $widget ||
+			! isset( $wp_widget_factory->widgets[ $widget ] ) ||
+			! ( $wp_widget_factory->widgets[ $widget ] instanceof WP_Widget )
+		) {
+			return new WP_Error(
+				'widget_invalid',
+				__( 'Invalid widget.', 'gutenberg' ),
+				array(
+					'status' => 404,
+				)
+			);
+		}
+
+		$widget_obj = $wp_widget_factory->widgets[ $widget ];
+
+		$instance = $request->get_param( 'instance' );
```

Comment:
I'm confused: why do we need both `instance` and `instance_changes`?

Reply:
The instance represents, the previous widget instance normally saved on the database (in our case save as the block attributes). Instance changes represent the fields that are being changed. That's how the widgets work, the update method receives the previous instance and the changes and returns a new instance.